aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,722 @@
+    }
+
+    /// Returns the number of aggregators that are used in the current context.
+    fn num_aggregators(&self) -> u128 {
+        self.aggregators.len() as u128
+    }
+
+    /// Creates and a new Aggregator with a given `id` and a `limit`. The value
+    /// of a new aggregator is always known, therefore it is created in a data
+    /// state, with a zero-initialized value.
+    fn create_new_aggregator(&mut self, id: AggregatorID, limit: u128) {
+        let aggregator = Aggregator {
+            value: 0,
+            state: AggregatorState::Data,
+            limit,
+        };
+        self.aggregators.insert(id, aggregator);
+        self.new_aggregators.insert(id);
+    }
+
+    /// If aggregator has been used in this context, it is removed. Otherwise,
+    /// it is marked for deletion.
+    fn remove_aggregator(&mut self, id: AggregatorID) {
+        // Aggregator no longer in use during this context: remove it.
+        self.aggregators.remove(&id);
+
+        if self.new_aggregators.contains(&id) {
+            // Aggregator has been created in the same context. Therefore, no
+            // side-effects.
+            self.new_aggregators.remove(&id);
+        } else {
+            // Otherwise, aggregator has been created somewhere else.
+            self.destroyed_aggregators.insert(id);
+        }
+    }
+}
+
+/// Represents a single aggregator change.
+#[derive(Debug)]
+pub enum AggregatorChange {
+    // A value should be written to storage.
+    Write(u128),
+    // A delta should be applied to storage.
+    Apply(DeltaOp),
+    // A value should be deleted from the storage.
+    Delete,
+}
+
+/// Represents changes made by all aggregators during this context. This change
+/// set can be converted into appropriate `WriteSet` and `DeltaChangeSet` by the
+/// user, e.g. VM session.
+pub struct AggregatorChangeSet {
+    pub changes: BTreeMap<AggregatorID, AggregatorChange>,
+}
+
+/// Native context that can be attached to VM `NativeContextExtensions`.
+///
+/// Note: table resolver is reused for fine-grained storage access.
+#[derive(Tid)]
+pub struct NativeAggregatorContext<'a> {
+    txn_hash: u128,
+    resolver: &'a dyn TableResolver,
+    aggregator_data: RefCell<AggregatorData>,
+}
+
+impl<'a> NativeAggregatorContext<'a> {
+    /// Creates a new instance of a native aggregator context. This must be
+    /// passed into VM session.
+    pub fn new(txn_hash: u128, resolver: &'a dyn TableResolver) -> Self {
+        Self {
+            txn_hash,
+            resolver,
+            aggregator_data: Default::default(),
+        }
+    }
+
+    /// Returns all changes made within this context (i.e. by a single
+    /// transaction).
+    pub fn into_change_set(self) -> AggregatorChangeSet {
+        let NativeAggregatorContext {
+            aggregator_data, ..
+        } = self;
+        let AggregatorData {
+            destroyed_aggregators,
+            aggregators,
+            ..
```

Comment:
I guess new_aggregators is only used to filter out created but destroyed ones in the same session

Reply:
Yes!